Add test_project, test_app and tests

seblat opened the pull request.

Commit:
Add test_project, test_app and tests

LilithWittmann commented:
@seblat it seems like that the tests are not executed via travis.  https://travis-ci.com/serioeseGmbH/serious-django-permissions

Commit:
Fix minor issues in tests

Commit:
Add end to end test

Commit:
Add GlobalLevelTests

Commit:
Add info about authoriztaion to README.rst

Commit:
Run tests in travis

Commit:
Install serious-django-permission in travis enviroment

seblat commented on test_project/test_app/tests.py:
```diff
@@ -0,0 +1,149 @@
+        self.assertFalse(self.unauthorized_user.has_perm(GlobalPermission))
+
+    def test_authorized_user_has_permission(self):
+        self.assertTrue(self.authorized_user.has_perm(GlobalPermission))
```
[This test is currently failing](https://travis-ci.com/serioeseGmbH/serious-django-permissions/builds/98599055). Yet, when checking the permissions of `self.authorized_user`, the global permission shows up:
```
/serious-django-permissions/test_project/test_app/tests.py(136)test_authorized_user_has_permission()
-> self.assertTrue(self.authorized_user.has_perm(GlobalPermission))
(Pdb) Permission.objects.filter(user=self.authorized_user)
<QuerySet [<Permission: auth | permission | Enables the user to write in any model>]>
```
Is this possibly a problem of `.has_perm()` not recognizing the `GlobalPermission` in the right way? @cobalamin

Commit:
fix(permissions): fix global permissions by actually implementing them

the code that previously existed didn't really make sense,
but we didn't notice because it was never called.

the solution we provide here is a bit hacky, by encoding what's usually
`app_label.codename` as `auth.app_label_and_codename`, where
`app_label_and_codename` is just app_label and codename concatenated together
with a dot. The reason for this is that we can't go around creating contenttypes
with fixed names as the previous code assumed, that API just doesn't exist in Django.
So we have to refer to one contenttype that encodes "globality" once, and encode
what's usually app_name and codename in the codename together.

Commit:
refactor(tests): explicitly refer to Permission class in restricted view, this is now possible

cobalamin commented:
The tests seem to be fixed now except for one case: Checking global permissions on the restricted view. I'm not sure what your assumptions were there, but it seems like you wrote those tests with the expectation that a global permission gives global access. The truth is that global permissions just model permissions that aren't attached to any specific model. They don't grant something like a superuser permission. If you want to test global permissions for restricted views too (good!), you should probably create a second restricted view that requires `GlobalPermission`.

Commit:
Fix GlobalPermissionTests an add Permission.user_has_perm tests

Closes #1

seblat commented:
> I'm not sure what your assumptions were there, but it seems like you wrote those tests with the expectation that a global permission gives global access.

Exactly :)

>  If you want to test global permissions for restricted views too (good!), you should probably create a second restricted view that requires `GlobalPermission`.

Check

seblat commented:
Taking a look at the missing coverage:
```
coverage report -m --skip-covered
Name                                                              Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------
serious_django_permissions/groups.py                                 20      2    90%   17, 28
serious_django_permissions/management/commands/create_groups.py      23      3    87%   28-30
serious_django_permissions/models.py                                  9      1    89%   7
serious_django_permissions/permissions.py                            49      4    92%   34, 38
```

- `groups.py `, `management/commands/create_groups.py` and `permissions.py`:
Those are exceptions. Whats our policy on testing them?
Edit: I've already discussed this with Lilith and am currently implementing the tests triggering the exceptions.

- `models.py`:
`GlobalPermissionManager.get_query_set()` does not get called. Is this stil necesarry after [fixing the global permissions](https://github.com/serioeseGmbH/serious-django-permissions/pull/3/commits/70394130ae2c4029a6298e0f9e9bdcebaede2e02)? The tests still pass, when `GlobalPermissionManager` and the `objects = GlobalPermissionManager()` call in `GlobalPermission` are removed.

Commit:
Add README.rst for test_project

Commit:
Add tests for exceptions

Commit:
fix(global_permissions): implement manager the correct way

Commit:
Merge branch 'seblat/tests' of github.com:serioeseGmbH/serious-django-permissions into seblat/tests

Commit:
fix(readme): Fix link in README.rst

this is .rst, not .md :)

Commit:
Merge branch 'seblat/tests' of github.com:serioeseGmbH/serious-django-permissions into seblat/tests

cobalamin commented:
> Edit: I've already discussed this with Lilith and am currently implementing the tests triggering the exceptions.
:+1: 
In general, policy on testing exceptions is with `.assertRaises()` and at least one positive and one negative test. From the docs on how to use `assertRaises`:

```
with self.assertRaises(SomeException) as cm:
    do_something()
```

The above should also allow an inline class definition inside the `with` block which should then fail, if (as an example) the class definition name does not end in "Permission".

----

> GlobalPermissionManager.get_query_set() does not get called. Is this stil necesarry after fixing the global permissions? The tests still pass, when GlobalPermissionManager and the objects = GlobalPermissionManager() call in GlobalPermission are removed.

Yes, it is, for usability reasons. How the global permissions are actually implemented is an ugly implementation detail, so it's good to have a queryset on the GlobalPermission model that contains only the GlobalPermission. You could just write a simple test that only `GlobalPermission`s are contained in `GlobalPermission.objects.all()` (after you've created at least one `GlobalPermission`, of course). That would fix that missing coverage and actually be a much needed test to verify the internal implementation works :)

*Further note*: Could you elaborate on how you created the coverage report (what happened before the `coverage report -m --skip-covered` call), and note that in the `test_project` README?
(where is the `test_project` README? maybe you forgot to push? :D)

Commit:
permissions.py: Fix missing description exception

The former condition doesnt trigger the exception since
`description` is defined in
`Permission(ABC, metaclass=PermissionMetaclass)`.
This caused `not hasattr(cls, 'description')` to return `False`.

Commit:
Fixup tests exceptions

Commit:
models.py: Fix GlobalPermissionManager()

get_query_set() was not executed via a get request. It had to be
changed to get_query(set) (see:
https://github.com/django/django/blob/master/django/db/models/manager.py#L139).
The lookup via __model returned `FieldError: Related Field got
invalid lookup: name.` It had to be changed to __model.

Commit:
Merge branch 'seblat/tests' of github.com:serioeseGmbH/serious-django-permissions into seblat/tests

seblat commented:
> ```
>with self.assertRaises(SomeException) as cm:
>     do_something()
> ```
> The above should also allow an inline class definition inside the `with` block which should then fail, if (as an example) the class definition name does not end in "Permission".

`self.asserRaises()` is awesome :D I've tried
```
def test_invalid_group_name(self):
        with self.assertRaises(ImproperlyConfigured) as e:
            class InvalidName(Group):
                permissions = [
                    RestrictedModelPermission
                ]
```
Yet this results in
```
FAIL: test_invalid_group_name (test_app.tests.GroupLevelTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../serious-django-permissions/test_project/test_app/tests.py", line 120, in test_invalid_group_name
    class InvalidName(Group):
AssertionError: ImproperlyConfigured not raised
```
Therefore [I´ve put the faulty group in a seperate module](https://github.com/serioeseGmbH/serious-django-permissions/pull/3/commits/e1ad6467f0159ac536b5ad90fdb689d24d31514f#diff-b455adb3f5e461d5a5babda686fb8348)

---

> Yes, it is, for usability reasons. How the global permissions are actually implemented is an ugly implementation detail, so it's good to have a queryset on the GlobalPermission model that contains only the GlobalPermission. You could just write a simple test that only `GlobalPermission`s are contained in `GlobalPermission.objects.all()` (after you've created at least one `GlobalPermission`, of course). That would fix that missing coverage and actually be a much needed test to verify the internal implementation works :)

:heavy_check_mark: 

 ---

> _Further note_: Could you elaborate on how you created the coverage report (what happened before the `coverage report -m --skip-covered` call), and note that in the `test_project` README?
> (where is the `test_project` README? maybe you forgot to push? :D)

I did not push, since I wasn't sure whether I will have to drop my "Remove GlobalPermissionManager" commit :D Anyway, I've created the coverage report via:
```
coverage run --source='.' test_project/manage.py test test_project
coverage report -m
```
[as stated in the updated and now pushed README.rst](https://github.com/serioeseGmbH/serious-django-permissions/pull/3/commits/1e37cb22025cb500d6e6c832ade132461c073571) :)

seblat commented on test_project/test_app/tests.py:
```diff
@@ -39,6 +42,15 @@ def test_create_global_permission(self):
+        del settings.DEFAULT_GROUPS_MODULE
+
+        create_groups.Command().handle()
+        #TODO: hide and test stdout: self.assertEquals("DEFAULT_GROUPS_MODULE setting is not set!", output)
```
Do we want create_groups.Command().handle() to simply print the error?

Or do we want it to throw an exception? For that [the exception in create_groups.Command().handle()](https://github.com/serioeseGmbH/serious-django-permissions/blob/master/serious_django_permissions/management/commands/create_groups.py#L26) would have to be changed to something like:
```
if not settings.DEFAULT_GROUPS_MODULE:
        raise AttributeError("DEFAULT_GROUPS_MODULE setting is not set!")


        else:
            groups_module = settings.DEFAULT_GROUPS_MODULE
```

cobalamin replied:
Raising an exception is better, especially for programmatically calling :) Feel free to change that.

seblat replied:
Changed in https://github.com/serioeseGmbH/serious-django-permissions/commit/3bf6054150e5fe1899fbdc9138fc8e2e9e189f38

cobalamin commented on test_project/test_app/tests.py:
```diff
@@ -114,6 +140,21 @@ def test_authorized_user_accessing_view(self):
+        with self.assertRaises(ImproperlyConfigured) as e:
+            from .groups_invalid_name import InvalidName
+        self.assertEqual("A Group class's name must end with 'Group'.",
+            str(e.exception)
```
I think `assertIn()` is generally the better choice for expected error strings, and picking the important substring. That way minor changes to the strings won't necessarily cause a test failure/rewrite. Here, it could be `self.assertIn("must end with 'Group'", str(e.exception))`.

seblat replied:
Changed in https://github.com/serioeseGmbH/serious-django-permissions/pull/3/commits/f3d8c7a18360e72ccc9f8428c5aad26d0005b03d :)

Commit:
refactor exception in create_groups.Command().handle()

Commit:
refactor tests to use assertIn

cobalamin commented:
Is this - in your eyes - ready to merge now? To me it looks good.
